Create a MCP for Scout

enriquesanchez-elastic opened the pull request:
This commit introduces the @kbn/scout-mcp package, which provides a Model Context Protocol (MCP) server to expose Scout's browser testing functionalities to AI assistants like Cursor. Key features include:

- Configuration for Cursor to interact with Kibana.
- Debugging tools for capturing screenshots, page snapshots, and current state during test failures.
- Integration of various tools for browser automation, authentication, and API services.

Additionally, updates to package.json and tsconfig files ensure proper linking and type definitions for the new package.

New files include:
- README.md, QUICK_START.md, and DEBUG_QUICK_REF.md for user guidance.
- Core implementation files for server, session management, and utility functions.

This enhancement aims to streamline the testing process and improve the overall developer experience when using Scout with Cursor.

## Summary

Summarize your PR. If it involves visual changes include a screenshot or gif.


### Checklist

Check the PR satisfies following conditions. 

Reviewers should verify this PR satisfies this list as well.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/src/platform/packages/shared/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This was checked for breaking HTTP API changes, and any breaking changes have been approved by the breaking-change committee. The `release_note:breaking` label should be applied in these situations.
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [ ] The PR  description includes the appropriate Release Notes section, and the correct `release_note:*` label is applied per the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
- [ ] Review the [backport guidelines](https://docs.google.com/document/d/1VyN5k91e5OVumlc0Gb9RPa3h1ewuPE705nRtioPiTvY/edit?usp=sharing) and apply applicable `backport:*` labels.

### Identify risks

Does this PR introduce any risks? For example, consider risks like hard to test bugs, performance regression, potential of data loss.

Describe the risk, its severity, and mitigation for each identified risk. Invite stakeholders and evaluate how to proceed before merging.

- [ ] [See some risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx)
- [ ] ...

enriquesanchez-elastic committed:
Add @kbn/scout-mcp package for enhanced debugging capabilities in Cursor

This commit introduces the @kbn/scout-mcp package, which provides a Model Context Protocol (MCP) server to expose Scout's browser testing functionalities to AI assistants like Cursor. Key features include:

- Configuration for Cursor to interact with Kibana.
- Debugging tools for capturing screenshots, page snapshots, and current state during test failures.
- Integration of various tools for browser automation, authentication, and API services.

Additionally, updates to package.json and tsconfig files ensure proper linking and type definitions for the new package.

New files include:
- README.md, QUICK_START.md, and DEBUG_QUICK_REF.md for user guidance.
- Core implementation files for server, session management, and utility functions.

This enhancement aims to streamline the testing process and improve the overall developer experience when using Scout with Cursor.

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/yarn_deduplicate'

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/generate codeowners'

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/356588)
* Commit: c081122b9c3b23c33d0320ce05e375fd46dc3068

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/356588#019a3476-5f12-4d99-abb8-3af5d8c6eea4)

cc @enriquesanchez-elastic

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356588","number":356588,"commit":"c081122b9c3b23c33d0320ce05e375fd46dc3068"}],"number":356588}
buildkite-pr-comment-->

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint_all_files --no-cache --fix'

enriquesanchez-elastic committed:
Remove outdated debugging documentation and quick start guides from the kbn-scout-mcp package. This cleanup enhances the package by eliminating redundant files that are no longer necessary for users, streamlining the documentation for better clarity and focus on current features.

enriquesanchez-elastic committed:
Merge branch 'main' into scout-mcp

kibanamachine committed:
Changes from node scripts/capture_oas_snapshot   --include-path /api/status   --include-path /api/alerting/rule/   --include-path /api/alerting/rules   --include-path /api/actions   --include-path /api/security/role   --include-path /api/spaces   --include-path /api/streams   --include-path /api/fleet   --include-path /api/saved_objects/_import   --include-path /api/saved_objects/_export   --include-path /api/maintenance_window   --include-path /api/agent_builder --update

kibanamachine committed:
Changes from node scripts/lint_ts_projects --fix

kibanamachine committed:
TO FIX: Run node 'scripts/yarn_deduplicate && yarn kbn bootstrap' locally, or add an exception to src/dev/yarn_deduplicate/index.ts and then commit the changes and push to your branch

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/356836)
* Commit: b27f2aa1e4bd8aaf61f1cfd99559e6aa5c543585

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/356836#019a35c7-37b8-48e2-ba59-46f10bda797b)

### History
* :broken_heart: [Build #356786](https://buildkite.com/elastic/kibana-pull-request/builds/356786) failed 80598cd24915cdc5f19280e7ffd4923b8d977a49
* :broken_heart: [Build #356754](https://buildkite.com/elastic/kibana-pull-request/builds/356754) failed 234ee8fde8e27775281dd917a365e0ecd4cfa85b
* :broken_heart: [Build #356602](https://buildkite.com/elastic/kibana-pull-request/builds/356602) failed fc640c8b008ccc3567edabf004db3d8b907e80bf
* :broken_heart: [Build #356588](https://buildkite.com/elastic/kibana-pull-request/builds/356588) failed c081122b9c3b23c33d0320ce05e375fd46dc3068

cc @enriquesanchez-elastic

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356836","number":356836,"commit":"b27f2aa1e4bd8aaf61f1cfd99559e6aa5c543585"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356786","number":356786,"commit":"80598cd24915cdc5f19280e7ffd4923b8d977a49"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356754","number":356754,"commit":"234ee8fde8e27775281dd917a365e0ecd4cfa85b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356602","number":356602,"commit":"fc640c8b008ccc3567edabf004db3d8b907e80bf"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356588","number":356588,"commit":"c081122b9c3b23c33d0320ce05e375fd46dc3068"}],"number":356836}
buildkite-pr-comment-->

enriquesanchez-elastic committed:
Refactor kbn-scout-mcp package by removing deprecated methods and parameters

This commit eliminates the `formatAccessibilityTree` method from the `ScoutSession` class and removes the `ref` parameter from the `ClickParams`, `TypeParams`, and `WaitForParams` interfaces. These changes streamline the codebase by removing outdated functionality and clarifying the required parameters for user interactions. Additionally, updates to the `scoutListApiServices` function reflect the current status of API tools, indicating they are not yet available in the MCP server.

enriquesanchez-elastic committed:
Add Jest configuration and tests for kbn-scout-mcp package

This commit introduces a Jest configuration file for the kbn-scout-mcp package, enabling unit testing capabilities. Additionally, it adds comprehensive test cases for configuration loading, validation, session management, and various utility functions. The tests ensure that the package behaves as expected under different scenarios, enhancing code reliability and maintainability.

kibanamachine committed:
Changes from node scripts/eslint_all_files --no-cache --fix

enriquesanchez-elastic committed:
Refactor kbn-scout-mcp package by removing unused dependencies and updating TypeScript configuration

This commit removes unused dependencies from the kbn-scout-mcp package's package.json, streamlining the dependency management. Additionally, it updates the TypeScript configuration to include Jest types, enhancing the testing capabilities. Minor adjustments in test files improve mock handling and ensure proper validation of URLs in security tests.

enriquesanchez-elastic committed:
Update browser tests in kbn-scout-mcp package to use specific data-test-subj attributes

This commit modifies the browser test cases to utilize specific data-test-subj attributes for locators, enhancing the accuracy of element targeting during tests. These changes improve the reliability of the tests by ensuring they interact with the correct elements, particularly in scenarios involving click failures and input handling.

elasticmachine committed:
Merge branch 'main' into scout-mcp

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/357562)
* Commit: 737a6fba331050146873da0361abfd5e725ba2cd

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/357562#019a492a-c59e-43e2-b03e-4e20e69caae8)

### History
* :broken_heart: [Build #357272](https://buildkite.com/elastic/kibana-pull-request/builds/357272) failed 96e10503964e1ce602c02cd75f2e29cf2b2059a3
* :broken_heart: [Build #357197](https://buildkite.com/elastic/kibana-pull-request/builds/357197) failed d0c47a4bcaf30bb998a0a4958931a36617096826
* :broken_heart: [Build #357175](https://buildkite.com/elastic/kibana-pull-request/builds/357175) failed 4141ea719fee99f90ee1dd3e421d6c4fb570ea44
* :broken_heart: [Build #356836](https://buildkite.com/elastic/kibana-pull-request/builds/356836) failed b27f2aa1e4bd8aaf61f1cfd99559e6aa5c543585
* :broken_heart: [Build #356786](https://buildkite.com/elastic/kibana-pull-request/builds/356786) failed 80598cd24915cdc5f19280e7ffd4923b8d977a49
* :broken_heart: [Build #356754](https://buildkite.com/elastic/kibana-pull-request/builds/356754) failed 234ee8fde8e27775281dd917a365e0ecd4cfa85b

cc @enriquesanchez-elastic

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/357562","number":357562,"commit":"737a6fba331050146873da0361abfd5e725ba2cd"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/357272","number":357272,"commit":"96e10503964e1ce602c02cd75f2e29cf2b2059a3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/357197","number":357197,"commit":"d0c47a4bcaf30bb998a0a4958931a36617096826"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/357175","number":357175,"commit":"4141ea719fee99f90ee1dd3e421d6c4fb570ea44"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356836","number":356836,"commit":"b27f2aa1e4bd8aaf61f1cfd99559e6aa5c543585"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356786","number":356786,"commit":"80598cd24915cdc5f19280e7ffd4923b8d977a49"}],"number":357562}
buildkite-pr-comment-->

enriquesanchez-elastic committed:
Enhance session tests in kbn-scout-mcp package to include elasticsearch role structure

This commit updates the session test cases to incorporate an elasticsearch role structure with an empty cluster array. Additionally, it refines the browser tests by removing unnecessary variable declarations, ensuring a cleaner and more efficient test setup. These changes improve the clarity and maintainability of the test code.

enriquesanchez-elastic committed:
Update kbn-scout-mcp package to configure default target URL for Kibana and Elasticsearch, enhance error handling, and streamline session management

This commit modifies the README to reflect the new default target URL for Kibana (http://localhost:5620) and Elasticsearch (http://localhost:9220). It also improves error handling in the session and authentication tools, ensuring more informative error messages. Additionally, unnecessary functions related to session cookie management have been removed to simplify the codebase. These changes enhance the usability and maintainability of the kbn-scout-mcp package.

enriquesanchez-elastic committed:
Refactor kbn-scout-mcp package to remove page object tools and enhance browser automation capabilities

This commit removes the page object tools from the kbn-scout-mcp package, streamlining the codebase and focusing on browser automation features. The README has been updated to reflect these changes, emphasizing the use of Scout's browser automation for UI interactions. Additionally, the session management has been improved by extending the Playwright Page with Scout-specific functionality, enhancing the overall usability and maintainability of the package.

enriquesanchez-elastic committed:
Update browser test to remove undefined parameter in app navigation expectation

This commit modifies the browser test case to ensure the `gotoApp` method is called with the correct parameters by removing the unnecessary `undefined` argument. This change enhances the clarity and accuracy of the test, ensuring it aligns with the intended functionality.

kibanamachine committed:
Changes from node scripts/eslint_all_files --no-cache --fix

enriquesanchez-elastic committed:
Enhance Kibana URL helper in kbn-scout-mcp package to support query parameters and hash values

This commit updates the `createKibanaUrlHelper` function to accept an options parameter, allowing the inclusion of query string parameters and hash values in the generated URLs. Additionally, the `PathOptions` interface has been introduced to define the structure of these options, improving type safety and clarity. The session management has also been refined to ensure that the current page is always initialized correctly, throwing an error if it fails to do so.

enriquesanchez-elastic committed:
Enhance kbn-scout-mcp package with new authentication methods and test generation tools

This commit introduces several improvements to the kbn-scout-mcp package, including new convenience methods for authentication: `scoutLoginAsAdmin`, `scoutLoginAsViewer`, and `scoutLoginAsPrivileged`. Additionally, it adds tools for generating tests from session actions, such as `scoutGenerateTestFile`, and enhances debugging capabilities with functions like `scoutGetConsoleLogs` and `scoutAnalyzeWaitFailure`. The README has been updated to reflect these changes, providing clearer guidance on authentication and test generation processes.

kibanamachine committed:
Changes from node scripts/capture_oas_snapshot   --include-path /api/status   --include-path /api/alerting/rule/   --include-path /api/alerting/rules   --include-path /api/actions   --include-path /api/security/role   --include-path /api/spaces   --include-path /api/streams   --include-path /api/fleet   --include-path /api/saved_objects/_import   --include-path /api/saved_objects/_export   --include-path /api/maintenance_window   --include-path /api/agent_builder --update

kibanamachine committed:
TO FIX: Run node 'scripts/yarn_deduplicate && yarn kbn bootstrap' locally, or add an exception to src/dev/yarn_deduplicate/index.ts and then commit the changes and push to your branch

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/362205)
* Commit: a01729d9e2fb381acb340e8f58f63a3a282494da

### Failed CI Steps
* [Linting](https://buildkite.com/elastic/kibana-pull-request/builds/362205#019a964c-3701-4950-9d1c-2ccd70f5446b)

### History
* :broken_heart: [Build #362183](https://buildkite.com/elastic/kibana-pull-request/builds/362183) failed 46a2688c2ec7dc0b670587313d885b7448c396fb
* :broken_heart: [Build #358176](https://buildkite.com/elastic/kibana-pull-request/builds/358176) failed 944507d5f4fbfda7751d22074b7d3b772e03223d
* :broken_heart: [Build #358150](https://buildkite.com/elastic/kibana-pull-request/builds/358150) failed 5658db5f4c96084cdad8866c256b8b66c85337e9
* :yellow_heart: [Build #357598](https://buildkite.com/elastic/kibana-pull-request/builds/357598) was flaky 792db5660c0e2b8fa4e2ce086300c16a70270e95
* :broken_heart: [Build #357562](https://buildkite.com/elastic/kibana-pull-request/builds/357562) failed 737a6fba331050146873da0361abfd5e725ba2cd

cc @enriquesanchez-elastic

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/362205","number":362205,"commit":"a01729d9e2fb381acb340e8f58f63a3a282494da"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/362183","number":362183,"commit":"46a2688c2ec7dc0b670587313d885b7448c396fb"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/358176","number":358176,"commit":"944507d5f4fbfda7751d22074b7d3b772e03223d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/358150","number":358150,"commit":"5658db5f4c96084cdad8866c256b8b66c85337e9"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/357598","number":357598,"commit":"792db5660c0e2b8fa4e2ce086300c16a70270e95"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/357562","number":357562,"commit":"737a6fba331050146873da0361abfd5e725ba2cd"}],"number":362205}
buildkite-pr-comment-->

enriquesanchez-elastic committed:
Improve suitability analysis in scoutAnalyzeCypressPatterns function to handle undefined data gracefully

This commit enhances the `scoutAnalyzeCypressPatterns` function by adding a check to ensure that `suitabilityAnalysis.data` is defined before accessing its properties. This change prevents potential runtime errors and improves the robustness of the analysis process, ensuring that warnings are only generated when appropriate data is available.

enriquesanchez-elastic committed:
Merge branch 'main' into scout-mcp
